compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -663,6 +663,7 @@ typeOfUpdate = \case
   UCreate tpl arg -> checkCreate tpl arg $> TUpdate (TContractId (TCon tpl))
   UCreateInterface iface arg -> checkCreateInterface iface arg $> TUpdate (TContractId (TCon iface))
   UExercise tpl choice cid arg -> typeOfExercise tpl choice cid arg
+  USoftExercise tpl choice cid arg -> typeOfExercise tpl choice cid arg
```

Comment:
here we want a stronger rule than `typeOfExercise` to forbid `softExercise` on templates with keys, compare `typeOfFetchTemplate` and `typeOfSoftFetchTemplate` in the equivalent scala file `daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala`. 

```suggestion
  USoftExercise tpl choice cid arg -> typeOfSoftExercise tpl choice cid arg
```

However, while pointing this out I noticed the Haskell typechecker _doesn't_ have the stronger rule for `softFetch`, I'll take care of it in another PR.

Reply:
this is the PR: https://github.com/digital-asset/daml/pull/16943

Reply:
I propose I leave all changes for _soft_ type checking to this future PR.